src/coreclr/jit/morph.cpp

```diff
@@ -13860,6 +13864,37 @@ GenTree* Compiler::fgOptimizeBitwiseAnd(GenTreeOp* andOp)
     return nullptr;
 }
 
+//------------------------------------------------------------------------
+// fgOptimizeBitwiseXor: optimizes the "xor" operation.
+//
+// Arguments:
+//   xorOp - the GT_XOR tree to optimize.
+//
+// Return Value:
+//   The optimized tree, currently always a local variable, in case any transformations
+//   were performed. Otherwise, "nullptr", guaranteeing no state change.
+//
+GenTree* Compiler::fgOptimizeBitwiseXor(GenTreeOp* xorOp)
```

Comment:
You could leave a `TODO-Cleanup: move the optimizations from "fgMorphSmpOpOptional" to this method.`.

Reply:
I moved the xor related code into this method and noticed that the condition of this branch looks odd to me.
https://github.com/dotnet/runtime/blob/34493120aa0f698287a75ca6cb621ca4d0252b7f/src/coreclr/jit/morph.cpp#L13901-L13909
Is it actually possible for a child node of XOR node to be a compare node? If I'm not wrong then it is a good candidate to be removed and extend the previous condition with `|| op2->IsIntegralConst(1)` since "x ^ -1 " = "~x" and "x ^ 1" = "~x". Correct me if I'm wrong :)

Reply:
> Is it actually possible for a child node of XOR node to be a compare node?

Why not?
```scala
ldarg 0
ldarg 1
ceq
ldc.i4 1
xor
```
Would produce a tree that this is looking for. Not sure how often this is hit in practice (you can remove it and see what SPMI tells, I would guess we'll see lots of such patterns in auto-generated IL in tests).

Reply:
> What do you think of "x ^ 1" = "~x"?

Are you suggesting to apply it for any `x`, not just compares? But that would not be correct, it only affects the lowest bit.

Reply:
All right. Thank you.